docs: reactivity (#3)
spryker/oryx: docs/reactivity/reactive-components.md

# Build reactive components with Oryx

## Integrate with domain logic

> Components can use domain logic to fetch data. Domain services provide a convenient API to request data, e.g. in case of loading product data, the Product Service can be leveraged:

Reviewer: But I'd say it's our strong recommendation to use domain controllers to fetch domain data.

Author: for sure, but we're building the story here.

Author: so I keep it as is

In these code snippet, an observable is assigned to the local `product$` field. This will not do anything, since you need to _subscribe_ to an observable. There's a convenient approach in Oryx to subscribe using a decorator:

```ts
export class ProductPriceComponent {
  @subscribe()
  protected product$ = this.productService.get({ sku: '123' }).pipe(
    tap((product) => {
      //do stuff here
    })
  );
}
```

The `subscribe` decorator will subscribe to the observable, but also unsubscribe when the component is destroyed. This will ensure that there's no leaking memory in the application.

Another missing piece here is how the observed data can be used and rendered inside the component. This is described in the next section.

## Update the DOM

While the user navigates through pages in a single page application experience, it's crucial for components to be updated accordingly. This section will explain how data updates are propagated throughout the user interface in Oryx.

> RxJS conveniently operates on data streams and updates them in memory, but it won't synchronized this to the UI automatically. Each JavaScript framework ships its own opinionated method to update the DOM. The method of choice contributes significantly to the performance and user experience of the application.

Reviewer 2: ```suggestion
RxJS conveniently operates on data streams and updates them in memory, but it won't synchronize this to the UI automatically. Each JavaScript framework ships its own opinionated method to update the DOM. The method of choice contributes significantly to the performance and user experience of the application.
```

> The components provided in the Oryx libraries are build with LIT. LIT provides a highly efficient system to only synchronizes the minimum required updates to the DOM. When updates are loaded asynchronous, we need, however, to update the UI whenever new data is emitted. To make this as transparent as possible, a decorator (`@asyncState()`) is provided that can be used in the UI components. The decorator uses the `AsyncStateController` under the hood, which will request updates to the view whenever needed.

Reviewer 2: ```suggestion
The components provided in the Oryx libraries are build with Lit. Lit provides a highly efficient system to only synchronizes the minimum required updates to the DOM. When updates are loaded asynchronous, we need, however, to update the UI whenever new data is emitted. To make this as transparent as possible, a decorator (`@asyncState()`) is provided that can be used in the UI components. The decorator uses the `AsyncStateController` under the hood, which will request updates to the view whenever needed.
```

The following example shows the use of the `asyncState` decorator. The decorator subscribes to the assigned observable and requests an update to the component whenever a needed. This means that as a component developer you do not need to worry about how the reactive system works under the hood.

> Oryx components are build in TypeScript, which is why we need to ensure type safety. The original type of the assigned observable needs to be adjusted. There's no TypeScript to resolve the type from the observable which is why the `valueType` function is offered to project the observed type.

Reviewer 2: ```suggestion
Oryx components are build in TypeScript, which is why we need to ensure type safety. The original type of the assigned observable needs to be adjusted. It's not possible to resolve correct TypeScript type from the observable by just using decorator, which is why the `valueType` function is offered to project the observed type.
```

    const salesPrice = this.priceService.format(price?.defaultPrice);
    const originalPrice = this.priceService.format(price?.originalPrice);
    return combineLatest({ salesPrice, originalPrice });
  }
}
```

**Note:** the example is intentionally simplified to keep the focus on the RxJs part.

In this example code you can see that the product data is observed from the ProductService, but switches to the price formatting logic. This means that whenever the product changes (i.e. on route change), new product data is emitted and formatted. The `PriceService` is used to format both the sales and original price. The `PriceService.format()` uses the current currency and locale for the formatting, which is why it also exposes an observable. Since there are two prices involved, the two streams are _combined_ in an object.

With this setup, the product price component will receive updated prices whenever the product, currency or locale in the application state will change.Subscribing to this code could simple, however more work is needed to ensure that the component UI (DOM) will be updated simultaneously. This is described in the next section.

## Ensuring compatibility with SSR

> Lorem impsem

Reviewer 2: ```suggestion
TODO
```
;)

# Reactive components

Components are organized by domains, for example components in a product domain, and can leverage domain logic to communicate with the associated backend API. Each domain is shipped with a domain service that provides an API to communicate with a backend API. For example, when rendering product data, `ProductService` can be used as follows:

```ts
export class ProductPriceComponent extends LitElement {
  protected productService = resolve(ProductService);
  protected product$ = this.productService.get({ sku: '123' });
}
```

> To ensure that components are reusable in different contexts, it is recommended to not couple them directly with the qualifier that is used to load data. For example, `ProductPriceComponent`does not need to be aware of the `SKU`, as the `SKU` can be determined from the route on the product page, the product card in a list, or the cart entry. Oryx provides a mechanism to setup a so-called context. In the case of product components, the product controller is used to resolve the SKU from the context controller.

Reviewer 3: What do you mean by the `route`?

Author: We can change it to URL actually. In frontend development we often speak about _routing_, which is the mechanism to reflect the application state by the URL. The route however does not include the domain name, protocol, etc. which is why we typically do not talk about the URL.

RxJS operates on data streams and updates them in memory, but it doesn't synchronize this to the UI automatically. Each JavaScript framework ships its own opinionated method to update the DOM. The method of choice contributes significantly to the performance and user experience of the application.

The components provided in the Oryx libraries are build with Lit. Lit provides a highly efficient system to only synchronize the minimum required updates to the DOM. When updates are loaded asynchronously, the UI needs to be updated every time new data is emitted. To make this as transparent as possible, the `@asyncState()` decorator can be used in the UI components. Under the hood, the decorator uses `AsyncStateController` which requests updates to the view when needed.

The following example shows the use of the `asyncState` decorator. The decorator subscribes to the assigned observable and requests an update to the component when needed. This means that as a component developer you do not need to worry about how the reactive system works under the hood.

> Oryx components are build in TypeScript, and we provides types everywhere to increase the developer experience and avoid error upfront. The original type of the assigned observable needs to be adjusted. It's impossible to resolve a correct TypeScript type from the observable by using a decorator, which is why the `valueType` function is offered to project the observed type.

Reviewer 3: `function is offered to project the observed type`. Is this correct? Looks a bit weird

Author: I mean it as a verb, from projection. we can rephrase.

Reviewer 3: If you can find a more simple, but still correct word, go for it.

Author: done

# Build reactive components with Oryx

## Integrate with domain logic

> Components can use domain logic to fetch data. Domain services provide a convenient API to request data, e.g. in case of loading product data, the Product Service can be leveraged:

Reviewer 3: Can components use anything else to fetch data or do they fetch data _using the domain logic_?

Author: yes and no. Using domain logic is recommended, but you could do differrently. 

we could add something like:

> Using the domain logic to interact with backend APIs is the recommended approach, but if you like to use alternative techniques to integrate with an API, nothing stops you from doing.

```ts
export class ProductPriceComponent extends LitElement {
  protected productService = resolve(ProductService);
  protected product$ = this.productService.get({ sku: '123' });
}
```

> The `ProductPriceComponent` should ideally not be aware of the SKU, as this is provided by the context; the SKU might be resolved from the URL or from the context, i.e. from a cart entry component. Oryx provides a context controller that can be used to resolve the SKU. Different domains have their own controller to resolve the right qualifier from the context. In case of the product domain we use the `ProductController`:

Reviewer 3: How is SKU being provided by the context ideal as opposed to the component being aware of it? What's wrong with the component being aware?

Author: I've improved the docs to better articulate why components should not be aware.

```ts
export class ProductPriceComponent extends LitElement {
  @property() sku?: string;
  protected productController = new ProductController(this);
  protected product$ = this.productController.get();
}
```

The `ProductController`resolves the product qualifier (SKU) from the context and returns an observable from the `ProductService`. It will also take a component property (`sku`) into account, in case the SKU is provided statically to the component. This might be useful in custom development or in case of demonstrating the component (e.g. Storybook).

> In these code snippet, an observable is assigned to the local `product$` field. This will not do anything, since you need to _subscribe_ to an observable. There's a convenient approach in Oryx to subscribe using a decorator:

Reviewer 3: What do you mean by `you need to _subscribe_ to an observable`?

Author: I've rewritten a little to better reflect this one.

```ts
export class ProductPriceComponent {
  @subscribe()
  protected product$ = this.productService.get({ sku: '123' }).pipe(
    tap((product) => {
      //do stuff here
    })
  );
}
```

The `subscribe` decorator will subscribe to the observable, but also unsubscribe when the component is destroyed. This will ensure that there's no leaking memory in the application.

Another missing piece here is how the observed data can be used and rendered inside the component. This is described in the next section.

## Update the DOM

While the user navigates through pages in a single page application experience, it's crucial for components to be updated accordingly. This section will explain how data updates are propagated throughout the user interface in Oryx.

RxJS conveniently operates on data streams and updates them in memory, but it won't synchronize this to the UI automatically. Each JavaScript framework ships its own opinionated method to update the DOM. The method of choice contributes significantly to the performance and user experience of the application.

> The components provided in the Oryx libraries are build with Lit. Lit provides a highly efficient system to only synchronizes the minimum required updates to the DOM. When updates are loaded asynchronous, we need, however, to update the UI whenever new data is emitted. To make this as transparent as possible, a decorator (`@asyncState()`) is provided that can be used in the UI components. The decorator uses the `AsyncStateController` under the hood, which will request updates to the view whenever needed.

Reviewer 3: Who/what provides the decorator?

Author: Oryx. I've updated the text

The following example shows the use of the `asyncState` decorator. The decorator subscribes to the assigned observable and requests an update to the component whenever a needed. This means that as a component developer you do not need to worry about how the reactive system works under the hood.

> Oryx components are build in TypeScript, which is why we need to ensure type safety. The original type of the assigned observable needs to be adjusted. It's not possible to resolve correct TypeScript type from the observable by just using decorator, which is why the `valueType` function is offered to project the observed type.

Reviewer 3: Is it the reader or Spryker who needs to ensure type safety?

Author: We are refactoring asyncState and valueType, so let's wait a little as this text will likely become obsolete.

  protected productService = resolve(ProductService);
  protected priceService = resolve(PriceService);

  protected prices = this.productService
    .getProduct()
    .pipe(switchMap((product) => this.formatPrices(product?.price)));

  protected formatPrices(
    price?: ProductPrices
  ): Observable<{ originalPrice: string | null; salesPrice: string | null }> {
    const salesPrice = this.priceService.format(price?.defaultPrice);
    const originalPrice = this.priceService.format(price?.originalPrice);
    return combineLatest({ salesPrice, originalPrice });
  }
}
```

**Note:** the example is intentionally simplified to keep the focus on the RxJs part.

In this example code you can see that the product data is observed from the ProductService, but switches to the price formatting logic. This means that whenever the product changes (i.e. on route change), new product data is emitted and formatted. The `PriceService` is used to format both the sales and original price. The `PriceService.format()` uses the current currency and locale for the formatting, which is why it also exposes an observable. Since there are two prices involved, the two streams are _combined_ in an object.

> With this setup, the product price component will receive updated prices whenever the product, currency or locale in the application state will change.Subscribing to this code could simple, however more work is needed to ensure that the component UI (DOM) will be updated simultaneously. This is described in the next section.

Reviewer 3: 1. Please doublecheck the second sentence, smth is amiss.
2. Where is the next section?

Author: has already been covered before in the text, have dropped the last section.
